patches/tModLoader/Terraria/WorldGen.TML.cs

```diff
@@ -7,4 +19,80 @@ public partial class WorldGen
 	private static double _timePass = 0.0; // Used to account for more precise time rates.
 
 	internal static void ClearGenerationPasses() => _generator?._passes.Clear();
+
+	internal static List<Tuple<GenPass, ILContext.Manipulator>> genPassILEdits;
+	internal static List<Tuple<GenPass, GenPassDetour>> genPassDetours;
+
+	public delegate void GenPassDetour(WorldGenLegacyMethod orig, GenerationProgress progress, GameConfiguration configuration);
+
+	/// <summary>
+	/// IL edits a PassLegacy during world generation. Use in ModifyWorldGenerationTasks or ModifyHardmodeTasks.
+	/// </summary>
+	/// <param name="task">The task to be IL edited</param>
+	/// <param name="callback">The method that IL edits the pass</param>
+	// TODO: Add support for hardmode tasks
+	public static void ModifyTask(GenPass task, ILContext.Manipulator callback)
+	{
+		HookEndpointManager.Modify(GetGenPassMethod(task), callback);
+		genPassILEdits.Add(Tuple.Create(task, callback));
```

Comment:
I don't like the idea of adding and removing IL hooks after load time. It 'should' be safe, but it's a fair bit of overhead, repeat logging, more risk of errors etc.

Fortunately, 1.4.4 made it so that none of the legacy gen pass delegates access any local variables anymore! That means you can do what we call a 'method split'

Break the `WorldGen.GenerateWorld` method into 3 methods, `GenerateWorld`, `AddGenPasses` and `GenerateWorld_RunTasksAndFinish`. 

```cs
		AddGenPasses();
		GenerateWorld_RunTasksAndFinish(seed, customProgressObject, generationStopwatch);
	}

	internal static void AddGenPasses() {
		AddGenerationPass("Reset", delegate (GenerationProgress progress, GameConfiguration passConfig) {
```

later
```cs
	private static void RunTasks(int seed, GenerationProgress customProgressObject, Stopwatch generationStopwatch) {
		SystemLoader.ModifyWorldGenTasks(_generator._passes, ref _generator._totalLoadWeight);
	
	...
```

Then you can call something like:
```cs
_generator = new();
AddGenPasses();
SystemLoader.HookGenPasses(_generator._passes)
```

Honestly, it doesn't look like any of the gen passes are conditionally added, so you could probably just store them all in a list at load, and expose it as `IReadOnlyList<GenPass> VanillaGenPasses` in `WorldGen.TML.cs`

Then your example could just use `WorldGen.EditGenPass("Living Trees", il=> ...`

Reply:
Would IL editing other mods generation passes be important? I'm guessing not, since it would probably be a smaller group of people wanting it, and I'm guessing most mods would use GenPasses instead of PassLegacys, but in case it is, should all world generation passes get moved to being added at load time (including modded ones)?

Reply:
IL editing requires intimate knowledge of the target code. Presumably, modded gen passes would have their code in one or more methods, not just a lambda.

Either way, I wouldn't bother considering mods IL editing mods when making tML decisions